### ISSUE TITLE
Add docstring to `torch/__init__.pyi`?

### ISSUE THREAD

**Issue by ShawnZhong** [pytorch/pytorch#37762]
## 🚀 Feature

TLDR: Could we attach docstring to `torch/__init__.pyi` so that VSCode/PyCharm/IntelliJ can show the documents in a popup window? 

It would make the development experiment much smoother if we could view the docs without leaving the IDE. 

## Motivation

We currently use a programmatically way (e.g., `add_docstr` as in [torch/_tensor_docs.py](https://github.com/pytorch/pytorch/blob/master/torch/_tensor_docs.py)) to set the docstring for functions written in C++/CUDA. 

While it works pretty well for Sphinx to generate the HTML docs, it makes the IDEs difficult to pick up the modified docstring without actually executing the code. 

Here is a simple code snippet to illustrate this point: 
| IDE | Screenshot |
| --- | --- | 
| VSCode | ![image](https://user-images.githubusercontent.com/6421097/80958455-adfa6580-8dca-11ea-8ef2-20cf20c206db.png) | 
| PyCharm <br> IntelliJ |  ![image](https://user-images.githubusercontent.com/6421097/80959326-375e6780-8dcc-11ea-8745-32eb12a7a563.png)| 

## Pitch

I am thinking if we could insert the docstring from [torch/_tensor_docs.py](https://github.com/pytorch/pytorch/blob/master/torch/_tensor_docs.py), [torch/_torch_docs.py](https://github.com/pytorch/pytorch/blob/master/torch/_torch_docs.py), [torch/_storage_docs.py](https://github.com/pytorch/pytorch/blob/master/torch/_storage_docs.py) to [torch/__init__.pyi](https://github.com/pytorch/pytorch/blob/master/torch/__init__.py) by modifying [tools/pyi/gen_pyi.py](https://github.com/pytorch/pytorch/blob/master/tools/pyi/gen_pyi.py). 

## Additional context

Here is a comparison between PyTorch and Tensorflow using `add` as an example. 

| IDE | `torch.add` | `tensorflow.add` | 
| --- | --- | --- | 
| VSCode | ![image](https://user-images.githubusercontent.com/6421097/80959800-29f5ad00-8dcd-11ea-8ede-a7c7c2a0ec60.png) | ![image](https://user-images.githubusercontent.com/6421097/80959833-3b3eb980-8dcd-11ea-8736-969e6dbcf4e8.png) | 
| PyCharm <br> IntelliJ | ![image](https://user-images.githubusercontent.com/6421097/80960836-48f53e80-8dcf-11ea-84f6-c0c6b9ba7db2.png) | ![image](https://user-images.githubusercontent.com/6421097/80960775-2105db00-8dcf-11ea-8adc-a5d0b57634ef.png) | 

cc @ezyang @zou3519

**Comment by ngimel** [comment]
cc @zou3519

**Comment by zou3519** [comment]
@ShawnZhong thank you for the report. We would definitely accept a PR to fix this.

It looks like if a stub file exists, then the IDE ignores the docstring :/. I did a quick [search](https://intellij-support.jetbrains.com/hc/en-us/community/posts/360000596824-Missing-docstrings-when-pyi-stub-present) on this and this might be a bug on the IDE's part, but it would be nice if it worked regardless of the IDE.

I'm not sure if modifying `tools/pyi/gen_pyi.py` is the way to go because that currently doesn't have a dependency on PyTorch and is used as a part of the build process. In order to grab the docstrings it looks like we might need to be able to `import torch`. I imagine if we go that route we'd have to refactor when gen_pyi gets run or add a postprocessing pass.

**Comment by zou3519** [comment]
Something related I found is that a lot of our "tensor" docstrings are not that good say "check the function variant". While this is OK on the web, if we're serious about making docstrings show up in IDEs then it might be good to flush them out a bit more:

![image](https://user-images.githubusercontent.com/5652049/81014937-ca7aba00-8e2b-11ea-8c1f-d9ff8992fe6d.png)

**Comment by ShawnZhong** [comment]
I'm not sure if it's the best way to do it, but it's possible to mock a `torch` module with `unittest.mock`: 

```py
import sys
from unittest.mock import Mock, patch

docstrings = {}


def gather_docstrings(func, docstr):
    docstrings[func._extract_mock_name()] = docstr


with patch.dict(sys.modules, {"torch": Mock(name="torch"), "torch._C": Mock(_add_docstr=gather_docstrings)}):
    sys.path.append("torch")  # bypassing torch/__init__.py
    import _torch_docs

print(docstrings)
# result: {'torch.abs': '\nabs(input, out=None) -> Tensor...', ...}
```

**Comment by ezyang** [comment]
Before you charge into writing a patch, I think there are some overlapping design considerations going on here.

In #16574 our long term goal is to make it possible to typecheck PyTorch codebase itself, and one subgoal of this is moving pyi annotations inline to Python files directly, rather than keeping them as separate stubs. What won't change, however, is that we will still need to generate stubs for `torch._C`, which doesn't have a Python source file at all.

So if we want to add docstrings to the pyi files, we should probably first restructure the scripts for the grand new world order.

This is on top of the problem of how to get the docstrings to the pyi file in the first place. Do docstrings in pyi show up in regular Python interpreter sessions? That's another consideration: we want `help(foo)` to keep working.

**Comment by ezyang** [comment]
Although it may be a lot of work, it might be correct for us to store our doc strings in some non-Python format, and then use codegen to make them available in Python and for IDEs.

**Comment by bonassifabio** [comment]
Is there any update on this proposed enhancement?
The poor docstrings displayed in IDEs are quite annoying, and force one to keep an interactive session open just to use help() or to use the browser, which can be a non-marginal waste of time.

**Comment by ezyang** [comment]
I stand by my suggestion at https://github.com/pytorch/pytorch/issues/37762#issuecomment-624364354 ; it would also help with better documenting Tensor.h on the C++ side c.f. this internal request https://fb.workplace.com/groups/894363187646754/posts/1360626894353712 (FB only); cc @jbschlosser @mruberry @albanD @ikriv

**Comment by sh-shahrokhi** [comment]
Hello, any news on this? It keeps me switching to chrome and back many times.

**Comment by ShawnZhong** [comment]
Since this issue hasn't been making progress for more than 3 years, I have attempted to implement my proposed solution in #114705 by including docstring to the `.pyi` files. Please let me know if you have any suggestions or feedbacks. 

To address some comments above, 

> It looks like if a stub file exists, then the IDE ignores the docstring :/.

> Do docstrings in pyi show up in regular Python interpreter sessions? That's another consideration: we want help(foo) to keep working.

No, docstrings in pyi do not show up w/ `help`. The current approach does not remove existing `__doc__`. Instead, it reads existing docstrings and attaches them to the `.pyi` files just to help the IDE. 

> I'm not sure if modifying tools/pyi/gen_pyi.py is the way to go because that currently doesn't have a dependency on PyTorch and is used as a part of the build process. In order to grab the docstrings it looks like we might need to be able to import torch. I imagine if we go that route we'd have to refactor when gen_pyi gets run or add a postprocessing pass.

Luckily, the docs module (e.g., [`_torch_docs.py`](https://github.com/pytorch/pytorch/blob/main/torch/_torch_docs.py)) has little dependencies for PyTorch, and can be imported painlessly by some mocking described before. 

> one subgoal of this is moving pyi annotations inline to Python files directly, rather than keeping them as separate stubs.

I agree. I don't see too many values to put the annotations (and docstrings) in `.pyi` instead of `.py`, given that such `.py` file exists (e.g., not generated from C sources) 

> What won't change, however, is that we will still need to generate stubs for torch._C, which doesn't have a Python source file at all.

That's true. This PR mainly focused on the docs for `torch._C`.

> Although it may be a lot of work, it might be correct for us to store our doc strings in some non-Python format, and then use codegen to make them available in Python and for IDEs.

I can see how this approach might work with a decent amount of work. I am looking at [_torch_docs.py](https://github.com/pytorch/pytorch/blob/main/torch/_torch_docs.py) and wonder if the docs are stored in some non-Python format (say YAML or something), how would the logic for string interpolation work (e.g., `common_args`, `factory_like_common_args` in the file). I could imagine using some YAML template and then having a python file to generate a real YAML file with string interpolated. The final YAML file can then be used to generate stubs and something else. Maybe I'm thinking too much, but that sounds complicated.

**Comment by ShawnZhong** [comment]
Since 2.3.0, docstrings are available on VSCode and PyCharm/IntelliJ!

<img width="869" alt="image" src="https://github.com/pytorch/pytorch/assets/6421097/5800d7c7-acf8-408d-9074-1cf57d40093a">

<img width="631" alt="image" src="https://github.com/pytorch/pytorch/assets/6421097/b522907a-d003-43f4-9544-ad0921ecc1cf">

**Comment by albanD** [comment]
Thanks for contributing this!
